fss: move detach logic to csi

adikul30 opened the pull request:
<!-- Thanks for sending a pull request! -->

**What this PR does / why we need it**:

For cases of ungraceful shutdown of nodes in supervisor environment, we need a way for CSI to detach volumes when spherelet cannot perform the detach. 

This code is being guarded by a supervisor capability which is currently disabled. 

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**Testing done**:

Tests with capability disabled

```
Build #1455 (Jun 26, 2024, 7:40:41 PM)
adkulkarni
PR 2695
FAIL! -- 12 Passed | 1 Failed | 0 Pending | 864 Skipped --- FAIL: TestE2E (3114.08s) FAIL Ginkgo ran 1 suite in 53m21.484744816s Test Suite Failed make: Leaving directory `/home/worker/workspace/csi-wcp-precheckin@2/Results/1455/vsphere-csi-driver'
```

```
Build #1456 (Jun 26, 2024, 8:45:03 PM)
adkulkarni
PR 2695
Ran 1 of 877 Specs in 785.871 seconds SUCCESS! -- 1 Passed | 0 Failed | 0 Pending | 876 Skipped PASS Ginkgo ran 1 suite in 14m41.702529624s Test Suite Passed make: Leaving directory `/home/worker/workspace/csi-wcp-precheckin@2/Results/1456/vsphere-csi-driver'
```

Sanity test with capability enabled ->

```
root@4214c7c56fd01429bba83088f0498336 [ ~ ]# k -n kube-system get cm wcp-cluster-capabilities -o yaml | grep CSI
  CSI_Detach_Supported: "true"
  
  
root@4214c7c56fd01429bba83088f0498336 [ ~ ]# k create -f podpvc.yaml -n adi-ns
persistentvolumeclaim/aditya-pvc created
pod/aditya-pod created

root@4214c7c56fd01429bba83088f0498336 [ ~ ]# k -n adi-ns get po
NAME         READY   STATUS    RESTARTS   AGE
aditya-pod   1/1     Running   0          33s
root@4214c7c56fd01429bba83088f0498336 [ ~ ]# k -n adi-ns get pvc
NAME         STATUS   VOLUME                                     CAPACITY   ACCESS MODES   STORAGECLASS                VOLUMEATTRIBUTESCLASS   AGE
aditya-pvc   Bound    pvc-3b4f13da-8f21-4e4c-baf8-0905c3117d85   1Gi        RWO            wcpglobal-storage-profile   <unset>                 35s

------

root@4214c7c56fd01429bba83088f0498336 [ ~ ]# k create -f sts.yaml -n adi-ns
service/nginx created
statefulset.apps/web created

root@4214c7c56fd01429bba83088f0498336 [ ~ ]# k -n adi-ns get sts
NAME   READY   AGE
web    1/1     117s

root@4214c7c56fd01429bba83088f0498336 [ ~ ]# k -n adi-ns scale sts web --replicas=10
statefulset.apps/web scaled

root@4214c7c56fd01429bba83088f0498336 [ ~ ]# k -n adi-ns get sts
NAME   READY   AGE
web    10/10   3m39s

root@4214c7c56fd01429bba83088f0498336 [ ~ ]# k -n adi-ns scale sts web --replicas=2
statefulset.apps/web scaled

root@4214c7c56fd01429bba83088f0498336 [ ~ ]# k -n adi-ns get sts
NAME   READY   AGE
web    2/2     6m6s

```


**Special notes for your reviewer**:

**Release note**:
<!--  Steps to write your release note:
1. Use the release-note-* labels to set the release note state (if you have access)
2. Enter your extended release note in the below block; leaving it blank means using the PR title as the release note. If no release note is required, just write `NONE`.
-->
```release-note
```

svcbot-qecnsdp commented:
Started WCP block pre-checkin pipeline... Build Number: 1178

svcbot-qecnsdp commented:
``` 
Build ID: 1178
WCP build status: FAILURE 
Stage before exit: null 

```

svcbot-qecnsdp commented:
Started WCP block pre-checkin pipeline... Build Number: 1180

svcbot-qecnsdp commented:
``` 
Build ID: 1180
WCP build status: FAILURE 
Stage before exit: null 
Jenkins E2E Test Results: 
Ran 11 of 809 Specs in 2690.181 seconds
FAIL! - Interrupted by User -- 9 Passed | 2 Failed | 0 Pending | 798 Skipped
--- FAIL: TestE2E (2690.35s)
FAIL

Ginkgo ran 1 suite in 45m57.023656047s

Test Suite Failed

```

svcbot-qecnsdp commented:
Started WCP block pre-checkin pipeline... Build Number: 1184

svcbot-qecnsdp commented:
``` 
Build ID: 1184
WCP build status: FAILURE 
Stage before exit: null 
Jenkins E2E Test Results: 
FAIL! -- 9 Passed | 4 Failed | 0 Pending | 796 Skipped
--- FAIL: TestE2E (3735.52s)
FAIL

Ginkgo ran 1 suite in 1h3m24.097574593s

Test Suite Failed
make: Leaving directory `/home/worker/workspace/csi-wcp-precheckin/Results/1184/vsphere-csi-driver`

```

svcbot-qecnsdp commented:
Started WCP block pre-checkin pipeline... Build Number: 1188

svcbot-qecnsdp commented:
``` 
Build ID: 1188
WCP build status: FAILURE 
Stage before exit: null 
Jenkins E2E Test Results: 
FAIL! -- 8 Passed | 5 Failed | 0 Pending | 796 Skipped
--- FAIL: TestE2E (3681.02s)
FAIL

Ginkgo ran 1 suite in 1h2m37.83448411s

Test Suite Failed
make: Leaving directory `/home/worker/workspace/csi-wcp-precheckin/Results/1188/vsphere-csi-driver`

```

svcbot-qecnsdp commented:
Started WCP block pre-checkin pipeline... Build Number: 1190

svcbot-qecnsdp commented:
``` 
Build ID: 1190
WCP build status: FAILURE 
Stage before exit: null 
Jenkins E2E Test Results: 
FAIL! -- 8 Passed | 5 Failed | 0 Pending | 796 Skipped
--- FAIL: TestE2E (3993.78s)
FAIL

Ginkgo ran 1 suite in 1h7m47.370367043s

Test Suite Failed
make: Leaving directory `/home/worker/workspace/csi-wcp-precheckin/Results/1190/vsphere-csi-driver`

```

svcbot-qecnsdp commented:
Started WCP block pre-checkin pipeline... Build Number: 1191

svcbot-qecnsdp commented:
``` 
Build ID: 1191
WCP build status: FAILURE 
Stage before exit: null 
Jenkins E2E Test Results: 
FAIL! -- 6 Passed | 7 Failed | 0 Pending | 796 Skipped
--- FAIL: TestE2E (5663.50s)
FAIL

Ginkgo ran 1 suite in 1h35m38.19888315s

Test Suite Failed
make: Leaving directory `/home/worker/workspace/csi-wcp-precheckin/Results/1191/vsphere-csi-driver`

```

svcbot-qecnsdp commented:
Started WCP block pre-checkin pipeline... Build Number: 1192

svcbot-qecnsdp commented:
``` 
Build ID: 1192
WCP build status: FAILURE 
Stage before exit: null 
Jenkins E2E Test Results: 
FAIL! -- 11 Passed | 2 Failed | 0 Pending | 796 Skipped
--- FAIL: TestE2E (3439.42s)
FAIL

Ginkgo ran 1 suite in 58m31.946189243s

Test Suite Failed
make: Leaving directory `/home/worker/workspace/csi-wcp-precheckin/Results/1192/vsphere-csi-driver`

```

svcbot-qecnsdp commented:
Started WCP block pre-checkin pipeline... Build Number: 1193

svcbot-qecnsdp commented:
``` 
Build ID: 1193
WCP build status: FAILURE 
Stage before exit: null 
Jenkins E2E Test Results: 
FAIL! -- 8 Passed | 5 Failed | 0 Pending | 796 Skipped
--- FAIL: TestE2E (3172.95s)
FAIL

Ginkgo ran 1 suite in 54m9.17423862s

Test Suite Failed
make: Leaving directory `/home/worker/workspace/csi-wcp-precheckin/Results/1193/vsphere-csi-driver`

```

svcbot-qecnsdp commented:
Started WCP block pre-checkin pipeline... Build Number: 1195

svcbot-qecnsdp commented:
``` 
Build ID: 1195
WCP build status: SUCCESS 
Stage before exit: finally 
Jenkins E2E Test Results: 

Ran 1 of 809 Specs in 402.654 seconds
SUCCESS! -- 1 Passed | 0 Failed | 0 Pending | 808 Skipped
PASS

Ginkgo ran 1 suite in 7m49.564279041s
Test Suite Passed
make: Leaving directory `/home/worker/workspace/csi-wcp-precheckin/Results/1195/vsphere-csi-driver`

```

adikul30 commented:
/remove-lifecycle stale

adikul30 commented:
@gohilankit, @divyenpatel can you help review this PR? thanks!

divyenpatel commented:
/ok-to-test

gohilankit commented on pkg/csi/service/wcp/controller.go:
```diff
@@ -1237,149 +1239,228 @@ func (c *controller) ControllerUnpublishVolume(ctx context.Context, req *csi.Con
+		if csiDetachEnabled {
+			var isVADeleted bool
+			var vmuuid string
+			volumeAttachment, err := commonco.ContainerOrchestratorUtility.GetVolumeAttachment(ctx, req.VolumeId, req.NodeId)
```
Will this check to verify that VA has been deleted go away sometime in the future? We don't do this check in other flavors like vanilla & pvcsi where CSI has always had the responsibility  to detach.

gohilankit commented on pkg/csi/service/wcp/controller.go:
```diff
@@ -1237,149 +1239,228 @@ func (c *controller) ControllerUnpublishVolume(ctx context.Context, req *csi.Con
+			}
+			return &csi.ControllerUnpublishVolumeResponse{}, "", nil
+		} else {
+			var isVADeleted bool
```
nit: Since it's the same logic we're invoking here and the case when WCP FSS is enabled, can we wrap getting VA object in a helper method and use that at both places?

adikul30 replied:
Since it was only ~10 lines, I ended up keeping it as is.

gohilankit commented:
I have one minor comment. Rest looks good.  
Can you update the PR with any testing done?

divyenpatel commented on pkg/csi/service/wcp/controller.go:
```diff
@@ -38,8 +38,9 @@ import (
 	"google.golang.org/protobuf/types/known/timestamppb"
 	v1 "k8s.io/api/core/v1"
-	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/errors"
```
it is better to use apierrors alias so we know we are using method from apimachinery.
We have similar error checks for VC API responses.

adikul30 replied:
Ok 👍

divyenpatel commented on pkg/csi/service/wcp/controller.go:
```diff
@@ -1237,149 +1239,228 @@ func (c *controller) ControllerUnpublishVolume(ctx context.Context, req *csi.Con
+			if !isVADeleted {
+				log.Debugf("controllerUnpublishVolumeInternal: VA : %v", spew.Sdump(volumeAttachment))
+				for k, v := range volumeAttachment.Status.AttachmentMetadata {
+					if k == common.AttributeVmUUID {
```
do we break once we found `k == common.AttributeVmUUID` and processed it.

adikul30 replied:
I can add a break. AttachmentMetadata is a map and `vmUUID` is the key we are looking for. There will only be 1 key like that in the map.

adikul30 commented:
I've added tests with capacity disabled. I'll also add a sanity test result with capacity enabled

adikul30 commented:
> I've added tests with capacity disabled. I'll also add a sanity test result with capability enabled

Added testing with statefulset in the PR description with capability enabled.

divyenpatel reviewed (approved):
/approve
/lgtm

divyenpatel commented:
/retest

divyenpatel commented:
/lgtm
